pubsub/api/Pubsub.Samples/PullMessagesWithCustomAttributesAsync.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START pubsub_subscriber_async_pull_custom_attributes]
+
+using Google.Cloud.PubSub.V1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+public class PullMessagesWithCustomAttributesAsyncSample
+{
+    public async Task<Dictionary<string, string>> PullMessagesWithCustomAttributesAsync(string projectId, string subscriptionId, bool acknowledge)
+    {
+        SubscriptionName subscriptionName = SubscriptionName.FromProjectSubscription(projectId, subscriptionId);
+
+        SubscriberClient subscriber = await SubscriberClient.CreateAsync(subscriptionName);
+        var attributes = new Dictionary<string, string>();
+        Task startTask = subscriber.StartAsync((PubsubMessage message, CancellationToken cancel) =>
+        {
+            string text = Encoding.UTF8.GetString(message.Data.ToArray());
+            Console.WriteLine($"Message {message.MessageId}: {text}");
+            if (message.Attributes != null)
+            {
+                foreach (var attribute in message.Attributes)
+                {
+                    attributes.Add(attribute.Key, attribute.Value);
+                    Console.WriteLine($"{attribute.Key} = {attribute.Value}");
+                }
+            }
+            return Task.FromResult(acknowledge ? SubscriberClient.Reply.Ack : SubscriberClient.Reply.Nack);
+        });
+        // Run for 7 seconds.
+        await Task.Delay(7000);
+        await subscriber.StopAsync(CancellationToken.None);
+        // Lets make sure that the start task finished successfully after the call to stop.
+        await startTask;
+        return attributes;
```

Comment:
I think it would be better to return the messages and on the test check that each message has the expected attributes.
It's weird to have a dictionary just contaning attributes without any reference to the message they are from, and also, it's very easy to make this code fail if several messages have the same attributes, our tests might not do that, but if a client copy-pasted this code to listen to their own messages, it will almost certainly happen.

Reply:
Just return list of message and change accordingly in test.